File: frontend/src/arkham/types/Phase.ts
import * as JsonDecoder from 'ts.data.json';

export type Phase = 'CampaignPhase' | 'InvestigationPhase' | 'EnemyPhase' | 'UpkeepPhase' | 'MythosPhase';

export const phaseDecoder = JsonDecoder.oneOf<Phase>([
  JsonDecoder.literal('CampaignPhase'),
  JsonDecoder.literal('InvestigationPhase'),
  JsonDecoder.literal('EnemyPhase'),
  JsonDecoder.literal('UpkeepPhase'),
  JsonDecoder.literal('MythosPhase'),
], 'Phase');

export type MythosPhaseStep
  = 'MythosPhaseBeginsStep'
  | 'PlaceDoomOnAgendaStep'
  | 'CheckDoomThresholdStep'
  | 'EachInvestigatorDrawsEncounterCardStep'
  | 'MythosPhaseWindow'
  | 'MythosPhaseEndsStep'

export type InvestigationPhaseStep
  = 'InvestigationPhaseBeginsStep'
  | 'InvestigationPhaseBeginsWindow'
  | 'NextInvestigatorsTurnBeginsStep'
  | 'NextInvestigatorsTurnBeginsWindow'
  | 'InvestigatorTakesActionStep'
  | 'InvestigatorsTurnEndsStep'
  | 'InvestigationPhaseEndsStep'

export type EnemyPhaseStep
  = 'EnemyPhaseBeginsStep'
  | 'HunterEnemiesMoveStep'
  | 'ResolveAttacksWindow'
  | 'ResolveAttacksStep'
  | 'AfterResolveAttacksWindow'
  | 'EnemyPhaseEndsStep'

export type UpkeepPhaseStep
  = 'UpkeepPhaseBeginsStep'
  | 'UpkeepPhaseBeginsWindow'
  | 'ResetActionsStep'
  | 'ReadyExhaustedStep'
  | 'DrawCardAndGainResourceStep'
  | 'CheckHandSizeStep'
  | 'UpkeepPhaseEndsStep'

export type PhaseStep = { tag: "MythosPhaseStep", contents: MythosPhaseStep }  | { tag: "InvestigationPhaseStep", contents: InvestigationPhaseStep }  | { tag: "EnemyPhaseStep", contents: EnemyPhaseStep }  | { tag: "UpkeepPhaseStep", contents: UpkeepPhaseStep }

const mythosPhaseStepDecoder = JsonDecoder.oneOf<MythosPhaseStep>([
  JsonDecoder.literal('MythosPhaseBeginsStep'),
  JsonDecoder.literal('PlaceDoomOnAgendaStep'),
  JsonDecoder.literal('CheckDoomThresholdStep'),
  JsonDecoder.literal('EachInvestigatorDrawsEncounterCardStep'),
  JsonDecoder.literal('MythosPhaseWindow'),
  JsonDecoder.literal('MythosPhaseEndsStep'),
], 'MythosPhaseStep');

const investigationPhaseStepDecoder = JsonDecoder.oneOf<InvestigationPhaseStep>([
  JsonDecoder.literal('InvestigationPhaseBeginsStep'),
  JsonDecoder.literal('InvestigationPhaseBeginsWindow'),
  JsonDecoder.literal('NextInvestigatorsTurnBeginsStep'),
  JsonDecoder.literal('NextInvestigatorsTurnBeginsWindow'),
  JsonDecoder.literal('InvestigatorTakesActionStep'),
  JsonDecoder.literal('InvestigatorsTurnEndsStep'),
  JsonDecoder.literal('InvestigationPhaseEndsStep'),
], 'InvestigationPhaseStep');

const enemyPhaseStepDecoder = JsonDecoder.oneOf<EnemyPhaseStep>([
  JsonDecoder.literal('EnemyPhaseBeginsStep'),
  JsonDecoder.literal('HunterEnemiesMoveStep'),
  JsonDecoder.literal('ResolveAttacksWindow'),
  JsonDecoder.literal('ResolveAttacksStep'),
  JsonDecoder.literal('AfterResolveAttacksWindow'),
  JsonDecoder.literal('EnemyPhaseEndsStep'),
], 'EnemyPhaseStep');

const upkeepPhaseStepDecoder = JsonDecoder.oneOf<UpkeepPhaseStep>([
  JsonDecoder.literal('UpkeepPhaseBeginsStep'),
  JsonDecoder.literal('UpkeepPhaseBeginsWindow'),
  JsonDecoder.literal('ResetActionsStep'),
  JsonDecoder.literal('ReadyExhaustedStep'),
  JsonDecoder.literal('DrawCardAndGainResourceStep'),
  JsonDecoder.literal('CheckHandSizeStep'),
  JsonDecoder.literal('UpkeepPhaseEndsStep'),
], 'UpkeepPhaseStep');

export const phaseStepDecoder = JsonDecoder.oneOf<PhaseStep>([
  JsonDecoder.object({ tag: JsonDecoder.literal('MythosPhaseStep'), contents: mythosPhaseStepDecoder }, 'MythosPhaseStep'),
  JsonDecoder.object({ tag: JsonDecoder.literal('InvestigationPhaseStep'), contents: investigationPhaseStepDecoder }, 'InvestigationPhaseStep'),
  JsonDecoder.object({ tag: JsonDecoder.literal('EnemyPhaseStep'), contents: enemyPhaseStepDecoder }, 'EnemyPhaseStep'),
  JsonDecoder.object({ tag: JsonDecoder.literal('UpkeepPhaseStep'), contents: upkeepPhaseStepDecoder }, 'UpkeepPhaseStep'),
], 'PhaseStep');

File: backend/arkham-core/library/Arkham/Phase.hs
{-# LANGUAGE TemplateHaskell #-}

module Arkham.Phase where

import Arkham.Prelude

import Data.Aeson.TH
import GHC.OverloadedLabels

data Phase
  = MythosPhase
  | InvestigationPhase
  | EnemyPhase
  | UpkeepPhase
  | ResolutionPhase
  | CampaignPhase
  deriving stock (Eq, Show, Ord, Data, Generic)
  deriving anyclass (NoThunks, NFData)

instance IsLabel "mythos" Phase where
  fromLabel = MythosPhase

instance IsLabel "investigation" Phase where
  fromLabel = InvestigationPhase

instance IsLabel "enemy" Phase where
  fromLabel = EnemyPhase

instance IsLabel "upkeep" Phase where
  fromLabel = UpkeepPhase

data PhaseStep
  = MythosPhaseStep MythosPhaseStep
  | InvestigationPhaseStep InvestigationPhaseStep
  | EnemyPhaseStep EnemyPhaseStep
  | UpkeepPhaseStep UpkeepPhaseStep
  deriving stock (Eq, Show, Ord, Data, Generic)
  deriving anyclass (NoThunks, NFData)

isMythosPhase :: Phase -> Bool
isMythosPhase MythosPhase = True
isMythosPhase _ = False

data MythosPhaseStep
  = MythosPhaseBeginsStep -- 1.1
  | PlaceDoomOnAgendaStep -- 1.2
  | CheckDoomThresholdStep -- 1.3
  | EachInvestigatorDrawsEncounterCardStep -- 1.4
  | MythosPhaseWindow -- fast player window
  | MythosPhaseEndsStep -- 1.5
  deriving stock (Eq, Show, Ord, Data, Generic)
  deriving anyclass (NoThunks, NFData)

data InvestigationPhaseStep
  = InvestigationPhaseBeginsStep -- 2.1
  | InvestigationPhaseBeginsWindow -- fast player window
  | NextInvestigatorsTurnBeginsStep -- 2.2
  | NextInvestigatorsTurnBeginsWindow -- fast player window
  | InvestigatorTakesActionStep -- 2.2.1
  | InvestigatorsTurnEndsStep -- 2.2.2
  | InvestigationPhaseEndsStep -- 2.3
  deriving stock (Eq, Show, Ord, Data, Generic)
  deriving anyclass (NoThunks, NFData)

data EnemyPhaseStep
  = EnemyPhaseBeginsStep -- 3.1
  | HunterEnemiesMoveStep -- 3.2
  | ResolveAttacksWindow -- fast player window
  | ResolveAttacksStep -- 3.3
  | AfterResolveAttacksWindow -- fast player window
  | EnemyPhaseEndsStep -- 3.4
  deriving stock (Eq, Show, Ord, Data, Generic)
  deriving anyclass (NoThunks, NFData)

data UpkeepPhaseStep
  = UpkeepPhaseBeginsStep -- 4.1
  | UpkeepPhaseBeginsWindow -- fast player window
  | ResetActionsStep -- 4.2
  | ReadyExhaustedStep -- 4.3
  | DrawCardAndGainResourceStep -- 4.4
  | CheckHandSizeStep -- 4.5
  | UpkeepPhaseEndsStep -- 4.6
  deriving stock (Eq, Show, Ord, Data, Generic)
  deriving anyclass (NoThunks, NFData)

$( do
    phase <- deriveJSON defaultOptions ''Phase
    phaseStep <- deriveJSON defaultOptions ''PhaseStep
    mythosPhaseStep <- deriveJSON defaultOptions ''MythosPhaseStep
    enemyPhaseStep <- deriveJSON defaultOptions ''EnemyPhaseStep
    investigationPhaseStep <- deriveJSON defaultOptions ''InvestigationPhaseStep
    upkeepPhaseStep <- deriveJSON defaultOptions ''UpkeepPhaseStep
    pure
      $ concat [phase, phaseStep, mythosPhaseStep, investigationPhaseStep, enemyPhaseStep, upkeepPhaseStep]
 )
